### ISSUE TITLE
genSymとinjectのプラグマを使い分けてexpanderにおけるimportのアクセス範囲を制御する

### ISSUE THREAD

**Issue by chaemon** [zer0-star/Nim-ACL#40]
現状ではexpanderはそのまま展開するので、ローカル変数に同名の変数やobjectが別のファイルにあるとエラーが起こるかと思います。したがって、ローカル変数のimportは全体に公開されないようにしたいと思っていました。つまり、expanderで書いたimportは事実上includeになっています。

前にzer0starさんがblockを使った方式を提案していただきましたが、expanderが関数名を全部把握しなくてはならない点や同名で引数だけが違う関数が使えないといった問題がありました。

これについていろいろ調べてみたところ、genSym, injectプラグマを利用してtemplateの形で展開すると良さそうだということがわかりました。現状、ローカル変数のバッティングがあまり起こらないので、expanderにわざわざその機能を入れる必要があるかというところではあります。

**Comment by chaemon** [comment]
つまり、expanderの方式をこれにすると、すべてのライブラリで以下の変更が必要になります。

- 変数・objectはデフォルトがgenSymで公開されないため、外に公開したい変数・objectすべてにgenSymをつける必要がある
- 関数(proc)はデフォルトがinjectで公開されるため、外に公開したくない関数(proc)すべてにinjectをつける必要がある

**Comment by zer0-star** [comment]
確かに、templateだと疑似的に名前空間のようなものを作れるのは良いかもしれませんね。とはいえ、大きな変更となるのでどうするか難しいところです。

**Comment by chaemon** [comment]
そうなんです。。。そもそもモジュール内にオブジェクトを隠したい場面があまり思いつかない割に、この変更を行うと既存のコードにinjectやgenSymをつけなくてはならなくなってしまうので、今の所必要ないかなーと思っています。

あるいは*がついたobject・変数に自動でinjectをつけて、*がついていないprocに自動でgenSymをつけるマクロを作りexpanderの中でそれが呼ばれるようにすれば自動で対応可能ではありますね。

**Comment by chaemon** [comment]
ところで、最近気づいたのですが、atcoderではコンパイル時にファイルを生成してそれをimportするみたいなことができちゃうようですねー

**Comment by zer0-star** [comment]
それ、めちゃくちゃ凄いですね......使えるかもしれません

**Comment by chaemon** [comment]
そうなんです！それ以外に、zip等のlinuxコマンドも呼び出せるので、さっそくNim-ACLを圧縮して送信し、それを展開して実行するexpanderの機能を作りました

**Comment by chaemon** [comment]
こんな感じに、コンパイル時にstaticExecでソースに圧縮して仕込んでおいたファイルを展開してatcoderというディレクトリを作って、そのディレクトリをそのままimportできました。

https://atcoder.jp/contests/abc256/submissions/33885705
